### ISSUE TITLE
Memory use

### ISSUE THREAD

**Issue by bibimainetti** [vmxdev/tkvdb#5]
Hello, I'm new to databases, I was trying your code, I noticed that it allocates a lot of memory (see tr_buf_allocated) when doing for example the test_iter(). It allocates something comparable to the database size itself. Is it the wanted behaviour?

**Comment by vmxdev** [comment]
Memory usage for radix trees heavy depends on data, e.g. keys (see Wikipedia articles for radix trees and tries descriptions). Random keys used in test is not a good choice for compact representation, keys with common prefixes are much better.

Internally each non-leaf node contains pointers to subnodes (tkvdb is 256-ary radix tree, so it's 1K for 32-bit addresses and 2K for 64-bit + 2K offsets in a file for non-RAM-only mode).

We "compress" sparse nodes before writing to disk in `commit()` function by removing empty pointers. But when you need it back (say, for iteration) affected nodes are read from disk in memory and becomes "uncompressed".

You can control size of RAM ("transaction") buffer by using DB or transaction parameter `TKVDB_PARAM_TR_LIMIT` - it sets the buffer size in bytes. When the buffer becomes full, you can still read from it, but all the operations that can grow up this buffer (reading additional nodes from disk, adding new k-v pair etc) will be rejected with TKVDB_ENOMEM.

So, in short - for some key sets radix trees can consume a lot of memory, but you don't need to keep all set in RAM and can use small enough buffer for adding key-value pairs or iterations (in cost of performance).

**Comment by bibimainetti** [comment]
Thank you very much for your explanation!
